services/src/main/java/org/keycloak/broker/saml/mappers/XPathAttributeMapper.java

```diff
@@ -0,0 +1,316 @@
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+        property = new ProviderConfigProperty();
+        property.setName(ATTRIBUTE_FRIENDLY_NAME);
+        property.setLabel("Friendly Name");
+        property.setHelpText("Friendly name of attribute to search for in assertion. You can leave this blank to try to apply XPath to all attributes or specify a name instead.");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+        property = new ProviderConfigProperty();
+        property.setName(USER_ATTRIBUTE);
+        property.setLabel("User Attribute Name");
+        property.setHelpText("User attribute name to store XPath value.  Use email, lastName, and firstName to map to those predefined user properties.");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        configProperties.add(property);
+    }
+
+    public static final String PROVIDER_ID = "saml-xpath-attribute-idp-mapper";
+
+    @Override
+    public boolean supportsSyncMode(IdentityProviderSyncMode syncMode) {
+        return IDENTITY_PROVIDER_SYNC_MODES.contains(syncMode);
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return configProperties;
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public String[] getCompatibleProviders() {
+        return COMPATIBLE_PROVIDERS;
+    }
+
+    @Override
+    public String getDisplayCategory() {
+        return "Attribute Importer";
+    }
+
+    @Override
+    public String getDisplayType() {
+        return "XPath Attribute Importer";
+    }
+
+    @Override
+    public void preprocessFederatedIdentity(KeycloakSession session, RealmModel realm, IdentityProviderMapperModel mapperModel, BrokeredIdentityContext context) {
+        String attribute = mapperModel.getConfig().get(USER_ATTRIBUTE);
+        if (StringUtil.isNullOrEmpty(attribute)) {
+            return;
+        }
+        String attributeName = getAttributeNameFromMapperModel(mapperModel);
+        String attributeXPath = mapperModel.getConfig().get(ATTRIBUTE_XPATH);
+
+        List<String> attributeValuesInContext = findAttributeValuesInContext(attributeName, attributeXPath, context);
+        if (!attributeValuesInContext.isEmpty()) {
+            if (attribute.equalsIgnoreCase(EMAIL)) {
+                setIfNotEmpty(context::setEmail, attributeValuesInContext);
+            } else if (attribute.equalsIgnoreCase(FIRST_NAME)) {
+                setIfNotEmpty(context::setFirstName, attributeValuesInContext);
+            } else if (attribute.equalsIgnoreCase(LAST_NAME)) {
+                setIfNotEmpty(context::setLastName, attributeValuesInContext);
+            } else {
+                context.setUserAttribute(attribute, attributeValuesInContext);
+            }
```

Comment:
Due to the same argument as above (https://issues.redhat.com/browse/KEYCLOAK-14536), can be replaced with 
```suggestion
            context.setUserAttribute(attribute, attributeValuesInContext);
```

Reply:
I think its correct to only call `setUserAttribute` when I find a value / the list is not empty

Reply:
That's right, and it it is on par with existing code (which alas contains still the specific methods for email / first name / lastname) https://github.com/keycloak/keycloak/blob/7e925bfbff5410d27675f10d596b752fa794a4ab/services/src/main/java/org/keycloak/broker/saml/mappers/UserAttributeMapper.java#L145-L155